### PR TITLE
[WIP] Following an object a certain distance

### PR DESCRIPTION
**Pull request by fmrico**
Signed-off-by: Francisco Martin Rico <fmrico@gmail.com>
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #1660  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | gazebo simulation of Tally |

---
This is a **work in progress**. I don't pretend to be accepted as is. I want only to discuss this approach.

The goal of this work is following a dynamic pose to a certain distance:
* The goal pose can change during the execution of `navigate_to_pose` action.
* The robot stops when the goal is nearer than a configured distance, and then it faces the goal.

One use case could be following a person to ~1 meter in a house. We want to use it in RoboCup@Home.

I don't want to implement this as a new plugin for controller or planner. This work is independent of the algorithms used to create the plan and following it. To implement it I have:

* developed a new control node in nav2_behavior_tree that overwrites the goal received in the `navigate_to_pose` action for the one received in the `goal_update` topic.
* added two input ports to `FollowPath` for indicating that never returns from the action, and the distance to the object. The other option was to create another action node in BT, action, action client, action server...

BT_carrot

This is the result (click for video):

[](https://www.youtube.com/watch?v=t7helr1IY_8&feature=youtu.be "Click to play on You Tube")

---

I don't know if this is the right place to discuss this, but the complete real use case I want to address is:
1. Make the robot navigate to the desired pose on the map.
2. Once there, make the robot follow a person (the dynamic goal pose, provided from outside Nav2)

The behavior trees used in the two steps in this application are (could be) different, but the BT to use is defined statically in the yaml param files. One option to solve this could be to select the BT to use in `NavigateToPose` action. Maybe `NavigateToPose` action could also include some options decided from behaviors, instead to be fixed in params.

Francisco

### CODE REVIEW

File: nav2_behavior_tree/README.md [hunk 1]
```diff
@@ -66,8 +66,9 @@ The nav2_behavior_tree package provides several navigation-specific nodes that a
 | BT Node   |      Type      |  Description |
 |----------|:-------------|------|
 | Backup |  Action | Invokes the BackUp ROS2 action server, which causes the robot to back up to a specific pose. This is used in nav2 Behavior Trees as a recovery behavior. The nav2_recoveries module implements the BackUp action server. |
-| ComputePathToPose |    Action   | Invokes the ComputePathToPose ROS2 action server, which is implemented by the nav2_planner module. The server address can be remapped using the `server_name` input port. |
 | FollowPath | Action |Invokes the FollowPath ROS2 action server, which is implemented by the controller plugin modules loaded. The server address can be remapped using the `server_name` input port. |
+| ComputePathToPose |    Action   | Invokes the ComputePathToPose ROS2 action server, which is implemented by the nav2_planner module. The server address can be remapped using the `server_name` input port. |
+| UpdateGoalPose | Action |Overwrite goal pose by one received in a topic. |
```
Comment (SteveMacenski) [inline nav2_behavior_tree/README.md:71]: Missing a space here between the pipe and Overwrite

File: nav2_behavior_tree/nav2_tree_nodes.xml [hunk 2]
```diff
@@ -42,6 +42,10 @@
       <input_port name="wait_duration">Wait time</input_port>
     </Action>
 
+    <Action ID="UpdateGoalPose">
+      <input_port name="goal">Destination to plan to</input_port>
```
Comment (SteveMacenski) [inline nav2_behavior_tree/nav2_tree_nodes.xml:46]:
```suggestion
      <input_port name="goal">Dynamically updating destination to plan to</input_port>
```

File: nav2_behavior_tree/CMakeLists.txt [hunk 3]
```diff
@@ -53,6 +53,9 @@ ament_target_dependencies(${library_name}
 add_library(nav2_compute_path_to_pose_action_bt_node SHARED plugins/action/compute_path_to_pose_action.cpp)
 list(APPEND plugin_libs nav2_compute_path_to_pose_action_bt_node)
 
+add_library(nav2_update_goal_pose_action_bt_node SHARED plugins/action/update_goal_pose_action.cpp)
```
Comment (SteveMacenski) [inline nav2_behavior_tree/CMakeLists.txt:56]: Can we add the word "dynamically" to the naming here to make it clear its for dynamic goal updating in a single task

File: nav2_behavior_tree/plugins/action/follow_path_action.cpp [hunk 4]
```diff
@@ -62,6 +64,8 @@ class FollowPathAction : public BtActionNode<nav2_msgs::action::FollowPath>
       {
         BT::InputPort<nav_msgs::msg::Path>("path", "Path to follow"),
         BT::InputPort<std::string>("controller_id", ""),
+        BT::InputPort<double>("distance_to_goal", "Distance to goal to stop translation"),
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/action/follow_path_action.cpp:67]: Shouldn't this distance to goal be computed by the dynamic new goal to follow (e.g. it is set at the distance away from the obstacle)

File: nav2_behavior_tree/plugins/action/update_goal_pose_action.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2018 Intel Corporation
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/action/update_goal_pose_action.cpp:1]: You should have your copyright here

File: nav2_behavior_tree/plugins/action/update_goal_pose_action.cpp [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  UpdateGoalPoseAction() = delete;
+
+  BT::NodeStatus tick() override
+  { 
+    if (!initialized_) {
+      initialize();
+    }
+
+    auto current_goal = config().blackboard->get<geometry_msgs::msg::PoseStamped>("goal");
+    if (rclcpp::Time(goal_.header.stamp) < rclcpp::Time(current_goal.header.stamp)) {
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/action/update_goal_pose_action.cpp:50]: Can't you directly compare stamps without the casting?

File: nav2_behavior_tree/plugins/action/update_goal_pose_action.cpp [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    config().blackboard->set<geometry_msgs::msg::PoseStamped>("goal", goal_);
+
+    return BT::NodeStatus::SUCCESS;
+  }
+
+  void initialize()
+  {
+    node_ = config().blackboard->get<rclcpp::Node::SharedPtr>("node");
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/action/update_goal_pose_action.cpp:61]: Can you remove the initialize method and add these to the constructor?

File: nav2_behavior_tree/plugins/action/update_goal_pose_action.cpp [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+    updated_goal_sub_ = node_->create_subscription<geometry_msgs::msg::PoseStamped>("goal_update",
+      10, std::bind(&UpdateGoalPoseAction::updatedGoalCallback, this, std::placeholders::_1));
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/action/update_goal_pose_action.cpp:64]: I think you really want a queue depth of 1 so that you always get the newest one

File: nav2_bt_navigator/behavior_trees/navigate_carrot.xml [hunk 9]
```diff
@@ -0,0 +1,17 @@
+<!--
+  This Behavior Tree replans the global path periodically at 1 Hz.
```
Comment (SteveMacenski) [inline nav2_bt_navigator/behavior_trees/navigate_carrot.xml:2]: You should update this description of the BT

File: nav2_bt_navigator/behavior_trees/navigate_carrot.xml [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+-->
+
+<root main_tree_to_execute="MainTree">
+  <BehaviorTree ID="MainTree">
+    <PipelineSequence name="NavigateWithReplanning">
+      <RateController hz="1.0">
+        <PipelineSequence name="GeneratePath">
+          <UpdateGoalPose goal="{goal}"/>
+          <ComputePathToPose goal="{goal}" path="{path}" planner_id="GridBased"/>
```
Comment (SteveMacenski) [inline nav2_bt_navigator/behavior_trees/navigate_carrot.xml:11]: Do you not also want a replanning rate in case your haven't gotten an updated dynamic goal pose recently?

File: nav2_controller/src/nav2_controller.cpp [hunk 11]
```diff
@@ -305,22 +306,70 @@ void ControllerServer::setPlannerPath(const nav_msgs::msg::Path & path)
     end_pose.pose.position.x, end_pose.pose.position.y);
 }
 
+double getDistance(
+  const geometry_msgs::msg::PoseStamped & pose_a,
+  const geometry_msgs::msg::PoseStamped & pose_b)
+{
+  double x_diff = fabs(pose_a.pose.position.x - pose_b.pose.position.x);
```
Comment (SteveMacenski) [inline nav2_controller/src/nav2_controller.cpp:313]: You don't need the `fabs` since you're throwing them into `hypot`

Comment (SteveMacenski) [reply nav2_controller/src/nav2_controller.cpp:313]: We also have a new utility in geometry_util in the nav2_util for this sort of thing, you should use it.

File: nav2_controller/src/nav2_controller.cpp [hunk 12]
```diff
@@ -305,22 +306,70 @@ void ControllerServer::setPlannerPath(const nav_msgs::msg::Path & path)
...
+  double y_diff = fabs(pose_a.pose.position.y - pose_b.pose.position.y);
+
+  return std::hypot(x_diff, y_diff);
+}
+
+geometry_msgs::msg::TwistStamped getTwistToPose(
```
Comment (SteveMacenski) [inline nav2_controller/src/nav2_controller.cpp:319]: Why is this here?

Comment (SteveMacenski) [reply nav2_controller/src/nav2_controller.cpp:319]: All of these changes in the controller server - I'm not sure they belong here. Can you explain this?

File: nav2_controller/src/nav2_controller.cpp [hunk 13]
```diff
@@ -305,22 +306,70 @@ void ControllerServer::setPlannerPath(const nav_msgs::msg::Path & path)
...
+  double distance_to_goal = action_server_->get_current_goal()->distance_to_goal;
+  bool keep_running = action_server_->get_current_goal()->keep_running;
 
   if (!getRobotPose(pose)) {
     throw nav2_core::PlannerException("Failed to obtain robot pose");
   }
 
-  progress_checker_->check(pose);
-
+  if (!keep_running) {
```
Comment (SteveMacenski) [inline nav2_controller/src/nav2_controller.cpp:358]: I'm not also sure about this keep running thing - if you've made it to `N` meters away from your goal, you would want it to stop. Once it starts moving again, then the controller can follow the new path to be `N` meters from that again.

Comment (SteveMacenski) [reply nav2_controller/src/nav2_controller.cpp:358]: I think the robot should slow down as it approaches the target so that you can keep smooth motion, else I think it would be pretty jerky

**Comment by SteveMacenski** [comment]
@fmrico what's the story with this PR?

**Comment by fmrico** [comment]
Hi @SteveMacenski, I have been very busy this week preparing the ROSDevDay2020, but I want to retake this work next week. I put this PR in stand by while I finished #1780, to provide a complete demo.

**Comment by SteveMacenski** [comment]
Look forward to your talk!

**Comment by fmrico** [comment]
Closed and continued in #1859
